prow/entrypoint/run_test.go

```diff
@@ -0,0 +1,106 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package entrypoint
+
+import (
+	"io/ioutil"
+	"os"
+	"path"
+	"testing"
+
+	"k8s.io/test-infra/prow/pod-utils/wrapper"
+)
+
+func TestOptions_Run(t *testing.T) {
+	var testCases = []struct {
+		name           string
+		args           []string
+		expectedLog    string
+		expectedMarker string
+		expectedErr    bool
+	}{
+		{
+			name:           "successful command",
+			args:           []string{"sh", "-c", "exit 0"},
+			expectedLog:    "",
+			expectedMarker: "0",
+			expectedErr:    false,
+		},
+		{
+			name:           "successful command with output",
+			args:           []string{"echo", "test"},
+			expectedLog:    "test\n",
+			expectedMarker: "0",
+			expectedErr:    false,
+		},
+		{
+			name:           "unsuccessful command",
+			args:           []string{"sh", "-c", "exit 12"},
+			expectedLog:    "",
+			expectedMarker: "12",
+			expectedErr:    true,
+		},
+		{
+			name:           "unsuccessful command with output",
+			args:           []string{"sh", "-c", "echo test && exit 12"},
+			expectedLog:    "test\n",
+			expectedMarker: "12",
+			expectedErr:    true,
+		},
+	}
+
+	for _, testCase := range testCases {
```

Comment:
Can we please include coverage of the timeout behavior?

Reply:
Do you have a suggestion of how to do this? We'd also have to refactor the timeout as right now it's an integer number of minutes and nobody wants to wait for that long in a test :)

Reply:
Thioughts on refactoring it to be a duration? At which point we could say 3s, 15m, etc (golang supports these values pretty readily)

Another idea would be:

```
var timeAfter = func(minutes int) chan time.Time { return time.After(minutes) }

func (o Options) Run() {
  ...
  end := timeAfter(timeoutMinutes)
  select {
  case err := <-done:
  case <-end:
}
```

and then in your test file:
```
func TestFoo() {
  old := timeAfter
  defer func() { timeAfter = old }()
  timeAfter = func(m int) chan time {
    c := make(chan time)
    go func() { c <- time.Now() }
    return c
  }
  o.Run()
}
```

If you want to retain the `t.Run()` concurrency you'll probably need to shove the `timeAfter chan` into `Options`

Reply:
Means we would need to refactor the API as you can't serialize to a `time.Duration`. I guess we could, though.

Reply:
Not necessary, the current form is fine. Just noting it would allow a quick sleep, although the `timeAfter` arg would allow no sleep

Reply:
I'd rather not complicate the normal flow only for the sake of testing -- the `timeAfter` thing immediately has someone asking "WTF" if they don't look at the test code